---
source: Rmd
title: Gene set analysis
teaching: XX
exercises: XX
---

```{r, echo=FALSE, purl=FALSE, message=FALSE}
source("download_data.R")
```

::::::::::::::::::::::::::::::::::::::: objectives

- Explain how to find differentially expressed pathways with gene set analysis in R.
- Understand how differentially expressed genes can enrich a gene set.
- Explain how to perform a gene set analysis in R, using clusterProfiler.

::::::::::::::::::::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::: questions

- How do we find differentially expressed pathways?

::::::::::::::::::::::::::::::::::::::::::::::::::

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("07-")
options(width = 120)
```

:::::::::::::::::::::::::::::::::::::::::  callout

### Contribute!

This episode is intended to introduce the concept of how to carry out a
functional analysis of a subset of differentially expressed (DE) genes,
by means of assessing how significantly DE genes enrich gene sets of
our interest.


::::::::::::::::::::::::::::::::::::::::::::::::::

First, we are going to explore the basic concept of enriching a gene set with
differentially expressed (DE) genes. Recall the differential expression analysis.

```{r, message=FALSE}
library(SummarizedExperiment)
library(DESeq2)
```

```{r}
se <- readRDS("data/GSE96870_se.rds")
```

```{r}
dds <- DESeq2::DESeqDataSet(se[, se$tissue == "Cerebellum"],
                            design = ~ sex + time)
```

```{r, message=FALSE}
dds <- DESeq2::DESeq(dds)
```

Fetch results for the contrast between male and female mice.

```{r}
resSex <- DESeq2::results(dds, contrast = c("sex", "Male", "Female"))
summary(resSex)
```

Select DE genes between males and females with FDR \< 5%.

```{r}
sexDE <- as.data.frame(subset(resSex, padj < 0.05))
dim(sexDE)
sexDE <- sexDE[order(abs(sexDE$log2FoldChange), decreasing=TRUE), ]
head(sexDE)
sexDEgenes <- rownames(sexDE)
head(sexDEgenes)
length(sexDEgenes)
```

## Enrichment of a curated gene set

:::::::::::::::::::::::::::::::::::::::::  callout

### Contribute!

Here we illustrate how to assess the enrichment of one gene set
we curate ourselves with our subset of DE genes with sex-specific
expression. Here we form such a gene set with genes from sex
chromosomes. Could you think of another more accurate gene set formed
by genes with sex-specific expression?


::::::::::::::::::::::::::::::::::::::::::::::::::

Build a gene set formed by genes located in the sex chromosomes X and Y.

```{r, message=FALSE}
xygenes <- rownames(se)[decode(seqnames(rowRanges(se)) %in% c("X", "Y"))]
length(xygenes)
```

Build a contingency table and conduct a one-tailed Fisher's exact test that
verifies the association between genes being DE between male and female mice
and being located in a sex chromosome.

```{r}
N <- nrow(se)
n <- length(sexDEgenes)
m <- length(xygenes)
k <- length(intersect(xygenes, sexDEgenes)) 
dnames <- list(GS=c("inside", "outside"), DE=c("yes", "no"))
t <- matrix(c(k, n-k, m-k, N+k-n-m),
                        nrow=2, ncol=2, dimnames=dnames)
t
fisher.test(t, alternative="greater")
```

## Gene ontology analysis with clusterProfiler

:::::::::::::::::::::::::::::::::::::::::  callout

### Contribute!

Here we illustrate how to assess the enrichment on the entire
collection of Gene Ontology (GO) gene sets using the package
clusterProfiler. Could we illustrate any missing important feature
of this package for this analysis objective? Could we briefly
mention other packages that may be useful for this task?


::::::::::::::::::::::::::::::::::::::::::::::::::

Second, let's perform a gene set analysis for an entire collection of gene sets
using the Bioconductor package
[clusterProfiler](https://bioconductor.org/packages/clusterProfiler). For this
purpose, we will fetch the results for the contrast between two time points.

```{r}
resTime <- DESeq2::results(dds, contrast = c("time", "Day8", "Day0"))
summary(resTime)
```

Select DE genes between `Day0` and `Day8` with FDR \< 5% and minimum 1.5-fold
change.

```{r}
timeDE <- as.data.frame(subset(resTime, padj < 0.05 & abs(log2FoldChange) > log2(1.5)))
dim(timeDE)
timeDE <- timeDE[order(abs(timeDE$log2FoldChange), decreasing=TRUE), ]
head(timeDE)
timeDEgenes <- rownames(timeDE)
head(timeDEgenes)
length(timeDEgenes)
```

Call the `enrichGO()` function from
[clusterProfiler](https://bioconductor.org/packages/clusterProfiler)
as follows.

```{r, message=FALSE}
library(clusterProfiler)
library(org.Mm.eg.db)
library(enrichplot)

resTimeGO <- enrichGO(gene = timeDEgenes,
                      keyType = "SYMBOL",
                      universe = rownames(se),
                      OrgDb = org.Mm.eg.db,
                      ont = "BP",
                      pvalueCutoff = 0.01,
                      qvalueCutoff = 0.01)
dim(resTimeGO)
head(resTimeGO)
```

Let's build a more readable table of results.

```{r}
library(kableExtra)

resTimeGOtab <- as.data.frame(resTimeGO)
resTimeGOtab$ID <- NULL
resTimeGOtab$geneID <- sapply(strsplit(resTimeGO$geneID, "/"), paste, collapse=", ")
ktab <- kable(resTimeGOtab, row.names=TRUE, caption="GO results for DE genes between time points.")
kable_styling(ktab, bootstrap_options=c("stripped", "hover", "responsive"), fixed_thead=TRUE)
```

:::::::::::::::::::::::::::::::::::::::: keypoints

- Key point 1

::::::::::::::::::::::::::::::::::::::::::::::::::


